circuit/src/gadgets/range.rs

```diff
@@ -0,0 +1,485 @@
+
+				ctx.next();
+
+				// Assign shifted value
+				let shifted_word = self.x.value().cloned() * Value::known(F::from(1 << (K - S)));
+				ctx.assign_advice(mock_common.common.advice[0], shifted_word)?;
+				ctx.enable(selector.clone())?;
+
+				ctx.next();
+
+				// Assign 2^{-S} from a fixed column.
+				let inv_two_pow_s = F::from(1 << S).invert().unwrap();
+				ctx.assign_from_constant(mock_common.common.advice[0], inv_two_pow_s)?;
+
+				Ok(assigned_x)
+			},
+		)
+	}
+}
+
+/// Range check chip using lookup table.
+///
+/// Check if the target value is `K * N` bits long.
+#[derive(Debug, Clone)]
+pub struct LookupRangeCheckChip<F: FieldExt, const K: usize, const N: usize> {
+	x: AssignedCell<F, F>,
+}
+
+impl<F: FieldExt, const K: usize, const N: usize> LookupRangeCheckChip<F, K, N> {
+	/// Construct a new instance
+	pub fn new(x: AssignedCell<F, F>) -> Self {
+		Self { x }
+	}
+}
+
+impl<F: FieldExt, const K: usize, const N: usize> MockChip<F> for LookupRangeCheckChip<F, K, N> {
+	type Output = AssignedCell<F, F>;
+
+	fn configure(
+		mock_common: &MockCommonConfig, meta: &mut halo2::plonk::ConstraintSystem<F>,
+	) -> Selector {
+		let running_sum_selector = meta.complex_selector();
+
+		let running_sum = mock_common.common.advice[0];
+
+		meta.lookup("running_sum check", |meta| {
+			let running_sum_selector = meta.query_selector(running_sum_selector);
+			let z_cur = meta.query_advice(running_sum, Rotation::cur());
+
+			// In the case of a running sum decomposition, we recover the word from
+			// the difference of the running sums:
+			//    z_i = 2^{K}⋅z_{i + 1} + a_i
+			// => a_i = z_i - 2^{K}⋅z_{i + 1}
+			let running_sum_lookup = {
+				let running_sum_word = {
+					let z_next = meta.query_advice(running_sum, Rotation::next());
+					z_cur.clone() - z_next * F::from(1 << K)
+				};
+
+				running_sum_selector.clone() * running_sum_word
+			};
+
+			// Combine the running sum and short lookups:
+			vec![(running_sum_lookup, mock_common.table)]
+		});
+
+		running_sum_selector
+	}
+
+	fn synthesize(
+		self, mock_common: &MockCommonConfig, selector: &Selector, mut layouter: impl Layouter<F>,
+	) -> Result<Self::Output, halo2::plonk::Error> {
+		layouter.assign_region(
+			|| "range check chip",
+			|region: Region<'_, F>| {
+				let mut ctx = RegionCtx::new(region, 0);
+				let x = ctx.copy_assign(mock_common.common.advice[0], self.x.clone())?;
+
+				let num_bits = K * N;
+
+				// Chunk the first num_bits bits into K-bit words.
+				let words = {
+					// Take first num_bits bits of `element`.
+					let bits = x.value().map(|element| {
+						fe_to_le_bits(element.clone())
+							.into_iter()
+							.take(num_bits)
+							.collect::<Vec<_>>()
+					});
+
+					bits.map(|bits| {
+						bits.chunks_exact(K)
+							.map(|word| F::from(lebs2ip::<K>(&(word.try_into().unwrap()))))
+							.collect::<Vec<_>>()
+					})
+					.transpose_vec(N)
+				};
```

Comment:
Also, can we avoid nested blocks

Reply:
Done https://github.com/eigen-trust/protocol/pull/169/commits/ab695818a4f077629984c7e35e847639235712ca